Make crypto:cipher_info work for all ciphers and aliases

essen opened the pull request:
Fix for https://bugs.erlang.org/browse/ERL-890

I'm not sure about the ASN1 type set to `undefined` for `aes_ige256`. It may have a type defined  in OpenSSL but I'm not sure where to find this information. Though I'm fine with `undefined` for now personally.

I'm also not sure about returning `ige_mode` for the cipher mode.

kostis commented on lib/crypto/test/crypto_SUITE.erl:
```diff
@@ -674,6 +674,7 @@ cipher_info(Config) when is_list(Config) ->
     #{type := _,key_length := _,iv_length := _,
         block_size := _,mode := _} = crypto:cipher_info(aes_128_cbc),
     {'EXIT',_} = (catch crypto:cipher_info(not_a_cipher)),
+    _ = [crypto:cipher_info(C) || C <- proplists:get_value(ciphers, crypto:supports())],
```
These comprehensions should be `lists:foreach/2` calls, really -- and then you can delete the following line.

essen replied:
Was a quick copy pasta from a larger snippet. I'll fix it, thanks.

essen commented:
I've added a second commit adding cipher modes that were missing. Seems I based my initial patch on OpenSSL 1.0 instead of 1.1. I added ifdefs to make sure it works with both but have only tested on OpenSSL 1.1 at the moment.

essen commented:
Updated the initial comment with a second concern I have about this PR. I'm not sure `ige_mode` makes sense there, although I've seen it described that way.

HansN commented:
Looks good in nightly tests, but aes_ctr, aes_{128,196,256}_ctr is not available with the EVP interface on older cryptolibs.  On them there is a notsup exception.
Maybe hard code like aes_ige or ar try-catch-end around the nif call and match on error:notsup.

essen commented:
I'm OK with hardcoding but what key size should be returned for `aes_ctr`? Same as `aes_256_ctr`?

HansN commented:
aes_ctr is a problem and has historic reasons.  And it is the same problem with aes_cbc, aes_cfb8, aes_cfb128, aes_ecb, aes_gcm and aes_ccm.  The aes_ige256 does also accept 128-bit keys, so there is also a problem. OpenSSL accepts 128, 196 and 256 for that one.
I would like to retire all of them including the aliases in crypto.erl, and only keep those with explicit key length,  but that could cause problems with compatibility.  So I'm sorry, but I don't know for now.

essen commented:
Yeah I agree with the sentiment. I will set 32 for now as this is what is returned when the cryptolib supports it at the moment, but that's completely arbitrary. Maybe this just needs a note in the documentation.

HansN commented:
Yes, do like that. I'll write some note in the documentation (which is not merged yet).

essen commented:
Pushed a commit with this addition.

HansN commented:
Thanks for the PR!

essen commented:
If I understand correctly you already got started on documentation changes? I can help review them when you're ready. If you didn't start I can work on them next week.
